File: TaskManager/src/services/CommandManager.java
package services;

import commands.Command;

import java.util.Stack;

public class CommandManager {
    private Stack<Command> commandStack = new Stack<>();
    private Stack<Command> redoStack = new Stack<>();

    public void executeCommand(Command command) {
        command.execute();
        commandStack.push(command);
        redoStack.clear();
    }

    public void undo() {
        if (!commandStack.isEmpty()) {
            Command command = commandStack.pop();
            command.undo();
            redoStack.push(command);
        }
    }

    public void redo() {
        if (!redoStack.isEmpty()) {
            Command command = redoStack.pop();
            command.execute();
            commandStack.push(command);
        }
    }

    public Stack<Command> getCommandHistory() {
        return commandStack;  // Expose the command history stack
    }
}

File: TaskManager/TaskManager/Commands/CommandManager.cs
using System.Collections.Generic;
namespace TaskManager.Commands;

public class CommandManager
{
    private readonly Stack<Command> _undoStack = new Stack<Command>();
    private readonly Stack<Command> _redoStack = new Stack<Command>();

    public void ExecuteCommand(Command command)
    {     command.Execute();
        _undoStack.Push(command);
        _redoStack.Clear();
    }
    
    public void Redo()
    {
        if (_redoStack.Count > 0)
        {
            var command = _redoStack.Pop();
            command.Execute();
            _undoStack.Push(command);
        }
    }
}